New chapter "Metropolis" with Python Implementation

shudipto-amin opened the pull request.

PeanutbutterWarrior commented:
I feel like this would make more sense grouped together with monte carlo under "Monte Carlo Methods", and rename monte carlo to "Approximating areas"

PeanutbutterWarrior commented:
I dislike using numpy. It's specialised enough that not everyone will know it, and it has no counterpart in many languages. It could be replaced here with `math.e^(-b1*(x-x1)**2)`, and `random.uniform(-1, 1)`.

shudipto-amin commented:
> I feel like this would make more sense grouped together with monte carlo under "Monte Carlo Methods", and rename monte carlo to "Approximating Areas"

I agree - they belong in a similar category.

henrikac commented on contents/metropolis/code/python/metropolis.py:
```diff
@@ -0,0 +1,54 @@
+import numpy as np
+import matplotlib.pyplot as pp
```
I'm not sure why `matplotlib` is imported because I don't see it used anywhere? :)

shudipto-amin replied:
My bad! I was plotting with it, but then forgot to remove it. That line can be deleted.

shudipto-amin commented:
> I dislike using numpy. It's specialised enough that not everyone will know it, and it has no counterpart in many languages. It could be replaced here with `math.e**(-b1*(x-x1)**2)`, and `random.uniform(-1, 1)`.

I'm not sure I agree, but I see where you're coming from, with `math` and `random` being built in libraries. Also, a bunch of other code on AAA uses numpy quite a bit. But perhaps here we don't need to use it, as it doesn't simplify the code like in the others. I am okay with changing it.

shudipto-amin commented on literature.bib:
```diff
@@ -346,3 +346,30 @@ @misc{3b1b_finger_count
+}
+
+}
+Equation of State Calculations by Fast Computing Machines
+Equation of State Calculations by Fast Computing Machines
```
Oops, these lines are unnecessary and should be deleted.

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,163 @@
+
+## Things to consider 
+
+### Any symmetric function can be used to generate the step
```
I think this entire paragraph is really important and should be moved up before the example code to where you talk about the definition of `g()`

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,163 @@
+So far the function $$g$$ we used for  generating the next step is a random number in the interval $$(-1,1)$$. However, this can be any function symmetric about $$0$$ for the above algorithm to work. For example, it can be a number randomly from a list of numbers like $$[ -3, -1, -1, +1, +1, +3]$$. In higher dimensions, the function should be symmetric in all directions, such as multidimensional  ...
+
+
+### A runaway walker
```
I don't know if I fully understand this section. Maybe a visualization would help?

Also: it might be a good idea to move this above the example code?

leios commented on contents/metropolis/code/python/metropolis.py:
```diff
@@ -0,0 +1,54 @@
+    xmin, xmax = -10, 10
+    x0 = np.random.uniform(xmin, xmax)
+
+    x_dat = [x0]  # for storing values
```
It might be nice to pre-allocate the array for performance (although I don't know python super well)

Amaras replied:
If performance really is an issue, I would recommend pre-allocating indeed.
I don't think this is really the case here, though.

shudipto-amin replied:
I did some research and  tried a few methods of pre-allocating - it turns out python lists aren't that efficient when pre-allocated, and appending turns out to be faster, surprisingly - or at best pre-allocating is about the same. Normally, for large iterations I believe we would just skip arrays altogether and just write directly to file after certain intervals.

Amaras replied:
I was thinking in terms of numpy arrays, rather than Python lists.
In the case of lists, the memory structure makes it hard to optimise by pre-allocation

shudipto-amin replied:
The main reason I didn't use numpy array in this case was because iterating over and changing a numpy array is actually slower than simply appending to a python list. I tried a few things, like using `np.nditer` and `np.ndenumerate`, but while those are fast at just reading the array, I couldn't make the process of re-assigning elements of an numpy array any faster. 

I think the main problem here is that the `metropolis_step()` function is not vectorizable. Eg. `y = some_function_of(x)`, is indeed vectorizable and much faster with numpy.  But what we have `x[ind] = some_function_of(x[ind -1])`, which is not vectorizable, and we have to explicitly loop through and assign elements of an ndarray - a slow process. What I usually do in this scenario is to use the package `numba`, which makes iterating over a pre-allocated numpy array super fast. But for the sake of this example, do we want to import more packages? Or is there some other trick that I'm unaware of?

Amaras replied:
> But for the sake of this example, do we want to import more packages? 

We only assume that the standard library, `numpy` and `matplotlib` are installed for Python, to have a somewhat standard, semi-minimalist, environment. I don't think we want to have `numba` installed as well.

> The main reason I didn't use numpy array in this case was because iterating over and changing a numpy array is actually slower than simply appending to a python list. I tried a few things, like using `np.nditer` and `np.ndenumerate`, but while those are fast at just reading the array, I couldn't make the process of re-assigning elements of an numpy array any faster.

Okay, fair enough. You might not want to not use `numpy.random` functions then, because it's probably not really worth it to import `numpy` in that case.
Maybe using a generator would be better here (rather than a list), but that can be confusing to beginners, so even though I'd recommend it in production code, I don't think it's necessary here.

shudipto-amin replied:
Might need to keep `numpy` after all - I realise now that the 'test' Leios is asking for requires calculating a histogram, which would be harder and slower to do without `numpy`.

leios commented on contents/metropolis/code/python/metropolis.py:
```diff
@@ -0,0 +1,54 @@
+    with open("output.dat", "w") as out:
+        out.write(output_string)
+        out.write("\n")
+        
```
It would be nice to have some sort of test to show that we are generating the correct function. An example test can be found in the [Approximate Counting Algorithm](https://www.algorithm-archive.org/contents/approximate_counting/approximate_counting.html)

shudipto-amin replied:
Done! It doesn't say "pass" or "fail", just tells the % error.

leios reviewed (commented):
Overall, I think the quality here is good. I a few lingering remarks that I didn't mention yet:

1. I think random walk should not be capitalized in the text body
2. It would be nice to talk briefly about how this is used in practice. I know it's used a lot in various physical applications, so maybe pointing to those with citations would be nice. I like that the code is for a general PDF, but maybe mentioning Mawell--Boltzmann or something would make the physics folk happy

leios commented:
For numpy, I think it's fine in this case. As someone coming from an HPC background, I genuinely thought the first thing everyone did when using python was to import numpy. I think we should be cognizant of the fact that this is not true for everyone and maybe remove numpy from less numerical methods.

Amaras commented on contents/metropolis/code/python/metropolis.py:
```diff
@@ -0,0 +1,54 @@
+    a2, b2, x2 =  3, 0.2, -1
+    a3, b3, x3 =  1,   2,  5
+    
+    return a1*np.exp( -b1*(x-x1)**2 ) + \
+           a2*np.exp( -b2*(x-x2)**2 ) + \
+           a3*np.exp( -b3*(x-x3)**2 )
```
PEP8 concern:
```suggestion
    return (
           a1 * np.exp(-b1 * (x - x1)**2) + 
           a2 * np.exp(-b2 * (x - x2)**2) + 
           a3 * np.exp(-b3 * (x - x3)**2)
        )
```

Amaras commented on contents/metropolis/code/python/metropolis.py:
```diff
@@ -0,0 +1,54 @@
+    '''Random step vector'''
+    return np.random.uniform(-1,1)
+
+def iterate(x, f=f, g=g):
```
The naming feels confusing to me. Shouldn't `iterate` do the complete loop?

shudipto-amin replied:
Hmm, would  `metropolis_step` or `one_complete_step` be better?

Amaras replied:
`metropolis_step` sounds better for me

Amaras commented on contents/metropolis/code/python/metropolis.py:
```diff
@@ -0,0 +1,54 @@
+    x_proposed = x + g()
+    A = min(1, f(x_proposed)/f(x) )
+    
+    u = np.random.uniform(0,1)
+    if u <= A:
+        x_new = x_proposed
+    else:
+        x_new = x
```
Fair enough, but wouldn't `np.random.choice` be a better function for this?, something like this:
```suggestion
x_new = np.random.choice([x_proposed, x], p=[A, 1-A])

shudipto-amin replied:
Hmm, yeah that probably makes it easier to understand too. The whole `if u <= A` code block is just how the algorithm is written in pseudo code. But it would be far less confusing to say "choose proposed x with probability A", and then use your suggestion. This way we can do away with the variable `u` completely.

Amaras commented on contents/metropolis/code/python/metropolis.py:
```diff
@@ -0,0 +1,54 @@
+    '''Perform one full iteration and return new position'''
+    
+    x_proposed = x + g()
+    A = min(1, f(x_proposed)/f(x) )
```
PEP8 recommends snake_case for variables in Python
```suggestion
    a = min(1, f(x_proposed) / f(x))
```

Amaras commented on contents/metropolis/code/python/metropolis.py:
```diff
@@ -0,0 +1,54 @@
+    x0 = np.random.uniform(xmin, xmax)
+
+    x_dat = [x0]  # for storing values
+    N = 50_000   # no. of iterations
```
Same problem as for `A`

shudipto-amin replied:
```suggestion
    num_steps = 50_000
```

Amaras reviewed (changes requested):
Alright, I don't have anything to add to the chapter review.
I have a few problems, though:
1. If you don't really use numpy features in your implementation, just don't import numpy. `math` functions are usually just as fast for non-vectorised calls, and I don't know much of the stdlib implementation of `random` to tell if `numpy.random` is faster or not.
2. You are technically not assured that `f` will not be 0, so that might get you into trouble

shudipto-amin commented on contents/metropolis/code/python/metropolis.py:
```diff
@@ -0,0 +1,54 @@
+    '''Random step vector'''
+    return np.random.uniform(-1,1)
+
+def iterate(x, f=f, g=g):
```
```suggestion
def metropolis_step(x, f=f, g=g):
```

shudipto-amin commented on contents/metropolis/code/python/metropolis.py:
```diff
@@ -0,0 +1,54 @@
+    x_proposed = x + g()
+    A = min(1, f(x_proposed)/f(x) )
+    
+    u = np.random.uniform(0,1)
+    if u <= A:
+        x_new = x_proposed
+    else:
+        x_new = x
```
```suggestion
    x_new = np.random.choice([x_proposed, x], p=[a, 1-a])
```

shudipto-amin commented on contents/metropolis/code/python/metropolis.py:
```diff
@@ -0,0 +1,54 @@
+    N = 50_000   # no. of iterations
+    
+    x = x0
+    for n in range(N):
```
```suggestion
    for n in range(num_steps):
```

shudipto-amin commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,163 @@
+P(\mathbf{x}) = \frac{f(\mathbf{x})}{\int_D f(\mathbf{x})d\mathbf{x}}
+$$
+
+where $$D$$ is the domain of $$P(\mathbf{x})$$, i.e., all possible values of the coordinates $$\mathbf{x}$$ for which $$P(\mathbf{x})$$ is defined. 
```
I think your comments are all good points, and I realize now that my introduction kind of blows. It is rushed, and assumes a lot of background from the reader. I think the way to make the intro better would be:

1. Add a subsection answering the question "What is a probability density function?" which would cover normalization and why normalization is difficult, with an example.
2. An "Applications" subsection which covers certain application areas like creating a Boltzmann distribution (or just distribution) of a system of particles. A good opportunity to cite more papers here too!

shudipto-amin commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,163 @@
+where $$D$$ is the domain of $$P(\mathbf{x})$$, i.e., all possible values of the coordinates $$\mathbf{x}$$ for which $$P(\mathbf{x})$$ is defined. 
+
+It's often easy for us to know $$f(x)$$. But the integral in the denominator can be quite difficult to calculate, even numerically.
+This is especially true when the coordinates ($$\mathbf{x}$$) are multidimensional, and $$f(\mathbf{x})$$ is an expensive calculation. One example use case is a multidimensional physical system with energy $$E(\mathbf{x})$$, for which we know that $$f(x) = e^{-E(\mathbf{x})}$$, but $$P(x)$$ is almost always impossible to calculate. 
```
I'm providing a Gaussian example, as well as an entire section for application :)

shudipto-amin commented:
Okay, I added a new section to Metropolis chapter on applications. I also created a new chapter on Probability Distributions which is linked to in the Metropolis chapter. Also added a bunch of citations and hopefully clarified a few points.

Amaras reviewed (changes requested):
Okay, so there are a few more corrections I'd like to see in the code from the current version. 

Some are not mandatory (transforming `metropolis_iterate` into a generator, using `numpy` arrays in `f`, and returning from `test_metropolis_iterate`), others like using the `%` formatting are definitely better than the current version.

I won't do a review of the probability distributions chapter right now, but I might do later. Leios will be the main chapter reviewer in any case, so...

shudipto-amin commented on contents/metropolis/code/python/metropolis.py:
```diff
@@ -0,0 +1,93 @@
+import numpy as np
+
+
+def f(x, normalize=False):
+    '''
+    Function proportional to target distribution, a sum of Gaussians.
+    For testing, set normalize to True, to get target distribution exactly.
+    '''
+    # Gaussian heights, width parameters, and mean positions respectively:
+    a1, b1, x1 = 10,   4, -4
+    a2, b2, x2 =  3, 0.2, -1
+    a3, b3, x3 =  1,   2,  5
+
+    if normalize:
+        norm = np.sqrt(np.pi)*(a1/np.sqrt(b1) + a2/np.sqrt(b2) + a3/np.sqrt(b3))
+        a1, a2, a3 = a1/norm, a2/norm, a3/norm
+
+    return (
+           a1 * np.exp(-b1 * (x - x1)**2) + 
+           a2 * np.exp(-b2 * (x - x2)**2) + 
+           a3 * np.exp(-b3 * (x - x3)**2)
+        )
```
I had to modify your code a bit to make it work (because the shape of `x` doesn't match that of `a`,`b`, or `xs` ):
```python
def f(x, normalize=False):
    '''
    Function proportional to target distribution, a sum of Gaussians.
    For testing, set normalize to True, to get target distribution exactly.
    '''
    # Gaussian heights, width parameters, and mean positions respectively:
    a  = np.array([10,   3, 1]).reshape(3, 1)
    b  = np.array([ 4, 0.2, 2]).reshape(3, 1)
    xs = np.array([-4,  -1, 5]).reshape(3, 1)

    if normalize:
        norm = np.sqrt(np.pi) * (a / np.sqrt(b)).sum()
        a /= norm
    return (a * np.exp(-b * (x - xs)**2)).sum(axis=0)
```
I'm okay with this if you are.

Amaras replied:
No problem with that, but please don't align numbers or operators (it's not PEP8 compliant)

shudipto-amin commented:
> Okay, so there are a few more corrections I'd like to see in the code from the current version.
> 
> Some are not mandatory (transforming `metropolis_iterate` into a generator, using `numpy` arrays in `f`, and returning from `test_metropolis_iterate`), others like using the `%` formatting are definitely better than the current version.
> 
> I won't do a review of the probability distributions chapter right now, but I might do later. Leios will be the main chapter reviewer in any case, so...

I believe the latest push addresses all of the above concerns. Also slighlty modified mardown to add a little discussion about the second video on metropolis walker.

Amaras reviewed (approved):
Okay, the Python part looks good to me :smiley:

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,152 @@
+
+Let's first define the domain of our target $$P(n)$$. We know that the lowest sum of two die is 2 (a 1 on both die), so $$n \geq 2$$ for sure. Similarly, the maximum is sum of two sixes, or 12, so $$n \leq 12$$ also. 
+
+So now we the domain of possibilites, i.e., $$n \in [2..12]$$. Next, we take a very common approach - we count up the number of different ways each of the possbile values of $$n$$ can occur. Let's call this the frequency, $$f(n)$$ of each possible $$n$$. We already know that $$f(2)=1$$, as there is only one way to get a pair of 1s. For $$n=3$$, we see that there are two possible ways: a $$1$$ and ...
```
```suggestion
So now we know the domain of possibilities, i.e., $$n \in [2..12]$$. Next, we take a very common approach - we count up the number of different ways each of the possbile values of $$n$$ can occur. Let's call this the frequency, $$f(n)$$ of each possible $$n$$. We already know that $$f(2)=1$$, as there is only one way to get a pair of 1s. For $$n=3$$, we see that there are two possible ways: a $$1$$ and $$2$$, or a $$2$$ and $$1$$, so $$f(3)=2$$. If you continue doing this for all $$n$$, you may see a pattern (homework for the reader!). Once you have all the $$f(n)$$, we can visualize it in a plot,
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,152 @@
+
+### Normalization 
+
+The $$f(n)$$ plotted above is technically NOT the probability $$P(n)$$ - because we know that the sum of all probabilities should be 1, which clearly isn't the case for $$f(n)$$. But we can just get that by dividing $$f(n)$$ by the _total_ number of possibilities, $$N$$. For two die, that is $$N = 6 \times 6 = 36$$, but we could also express it as the _sum of all frequencies_,
```
```suggestion
The $$f(n)$$ plotted above is technically NOT the probability $$P(n)$$ - because we know that the sum of all probabilities should be 1, which clearly isn't the case for $$f(n)$$. But we can get the probability by dividing $$f(n)$$ by the _total_ number of possibilities, $$N$$. For two die, that is $$N = 6 \times 6 = 36$$, but we could also express it as the _sum of all frequencies_,
```

leios replied:
I don't know if it's just me, but I have never called these frequencies and find that notation a bit confusing. I usually instead microstate / macrostate notation.

I think there at least needs to be an explicit mention that "frequency here means the number of occurrences of $$n$$ for each number." or something more elegant than that.

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,152 @@
+P(n) = \frac{f(n)}{\displaystyle\sum_{n} f(n)}
+$$
+
+Note that $$f(n)$$ does not necessarily have to be the frequency of $$n$$ - it could really be any function which is _proportional_ to $$P(n)$$, and the above definition of $$P(n)$$ would still hold. And it's easy to check that the sum is now equal to 1, since
```
```suggestion
Note that $$f(n)$$ does not necessarily have to be the frequency of $$n$$ - it could be any function which is _proportional_ to $$P(n)$$, and the above definition of $$P(n)$$ would still hold. And it's easy to check that the sum is now equal to 1, since
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,152 @@
+\sum_n P(n) = \frac{\displaystyle\sum_{n}f(n)}{\displaystyle\sum_{n} f(n)} = 1
+$$
+
+Once we have the probability function $$P(n)$$, we can calculate all sorts of probabilites. For example, let's say we want to find the probability that $$n$$ will be between two integers $$a$$ and $$b$$ inclusive. For brevity, we will use the notation $$\mathbb{P}(a \leq n \leq b)$$ to denote this probability. And to calculate it, we simply have to sum up all the probabilities for each value of $ ...
```
```suggestion
Once we have the probability function $$P(n)$$, we can calculate all sorts of probabilites. For example, let's say we want to find the probability that $$n$$ will be between two integers $$a$$ and $$b$$, inclusively (also including $$a$$ and $$b$$). For brevity, we will use the notation $$\mathbb{P}(a \leq n \leq b)$$ to denote this probability. And to calculate it, we simply have to sum up all the probabilities for each value of $$n$$ in that range, i.e.,
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,152 @@
+
+## Probability Density Functions
+
+What if instead of a discrete variable $$n$$, we had a continuous variable $$x$$, like temperature or weight? In that case, it doesn't make sense to ask what the probability is of $$x$$ being _exactly_ a particular number - there are infinite possible real numbers, after all, so the probability of $$x$$ being exactly any one of them is essentially zero! But it _does_ make sense to ask what the pr ...
```
```suggestion
What if instead of a discrete variable $$n$$, we had a continuous variable $$x$$, like temperature or weight? In that case, it doesn't make sense to ask what the probability is of $$x$$ being _exactly_ a particular number - there are infinite possible real numbers, after all, so the probability of $$x$$ being exactly any one of them is essentially zero! But it _does_ make sense to ask what the probability is that $$x$$ will be _between_ a certain range of values. For example, one might say that there is 50% chance that the temperature tomorrow at noon will be between 5 and 15, or 5% chance that it will be between 16 and 16.5. But how do we put all that information, for every possible range, in a single function? The answer is to use a __probability density function__. 
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,152 @@
+
+So strictly speaking, $$P(x)$$ itself is NOT a probability, but rather the probability is the quantity $$P(x)dx$$, or any area under the curve. That is why we call $$P(x)$$ the probability _density_ at $$x$$, while the actual probability is only defined for ranges of $$x$$. 
+
+But what about large ranges of $$x$$, which are not infinitesimally thin? We do exactly what we did for the discrete case - sum up the probabilities of each and every distinct range of values, each with an infinitesimal width $$dx$$. And what do we call such a sum over a continuous variable? Why, an integral, of course! Who knew calculus would come in handy one day? And so we have,
```
I feel this last line is out of place a bit. The discussion on discrete PDFs was almost at an undergrad level, while this feels more like high school? I don't know if we need to recap calculus or not?

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,152 @@
+For example, consider the __normal distribution function__, 
+
+$$
+f(x) = e^{-x^2}
```
You are right that this is the heart of a Gaussian, but I don't know if it's right to call this the normal distribution function without the standard deviation (for example). Maybe say that we "consider the normal distribution function, which boils down to..."

leios replied:
```suggestion
f(x) = e^{-x^2},
```

leios reviewed (commented):
Overall great chapter, I will get back to metropolis tomorrow. Would you be willing to put each sentence on it's own line? This helps with review and also maintainence.

shudipto-amin commented:
> Overall great chapter, I will get back to metropolis tomorrow. Would you be willing to put each sentence on it's own line? This helps with review and also maintainence.

I just saw how inconvenient this was for the review on your last stream - so sorry about this! I have modified both files (metropolis and probability chapters) to fix this.

shudipto-amin commented:
Just pushed the latest version of this PR, which addresses the recent comments and suggestions by @leios . In particular, this
1. Fixes the line separation issue for each sentence (both metropolis and PDF chapters).
2. Adds lines for implementing MathJax in PDF chapter, as indicated.
3. Adds intro to PDF. 
4. Removes intro to calculus in PDF chapter.
5. Makes discrete probability section more understandable (hopefully).
6. Makes other minor edits.

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,273 @@
+# Metropolis Algorithm
```
```suggestion
# The Metropolis Algorithm
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,273 @@
+# Metropolis Algorithm
+
+The [Monte Carlo Integration](../monte_carlo_integration/monte_carlo_integration.html) method uses random numbers to approximate the area of pretty much any shape we choose. 
+The Metropolis algorithm {{ "metropolis1953equation" | cite }} is a slightly more advanced Monte Carlo method which uses random numbers to approximate a [probability distribution](../probability/distributions/distributions.md).
+
+Here we will use the Metropolis algorithm to approximate a __probability density function__, as described in [this chapter](../probability/distributions/distributions.md), which is of the form,
+
+
```
```suggestion
The Metropolis algorithm {{ "metropolis1953equation" | cite }} is a slightly more advanced Monte Carlo method which uses random numbers to approximate a [probability distribution](../probability/distributions/distributions.md):

```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,273 @@
+$$
+
+where $$D$$ is the domain of $$P(\mathbf{x})$$, i.e., all possible values of the $$\mathbf{x}$$ for which $$P(\mathbf{x})$$ is defined; 
+and $$f(\mathbf{x})$$ is some a function that is proportional to $$P(x)$$, such as a frequency distribution of $$\mathbf{x}$$. 
```
Again, not super clear on what a "frequency distribution" is. Maybe saying "such as the likelihood of particular outcomes in a game of chance."

leios replied:
It might be nice to explicitly mention that the denominator is normalizing the function?

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,273 @@
+
+## An  example application
+
+ One example of a complicated probability function arises when considering a physical system of $$N$$ number of particles. 
```
```suggestion
One example of a complicated probability function arises when considering a physical system of $$N$$ particles. 
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,273 @@
+This is equivalent to dividing the length of a football field into intervals of about 2 meters - not a resolution you'd wanna watch a game in! 
+Even with such poor resolution, the number of different combinations of positions is $$10^{50}$$ - a colossal number indeed. 
+To see how large this number is, imagine that a single computation of $$E(\mathbf{x})$$ took only 1 nanosecond on a single processor, which is much faster than most energy calculations for physical systems in practice.
+ With that speed, it would require $$10^{41}$$ seconds on a single processor to calculate $$Q$$ - which means that _even_ with all the processors in the world running in parallel (there could be billions or trillions of them), calculating $$Q$$ would still take longer than the age of the universe - by many orders of magnitude!
```
I don't know if I buy this. We are giving each of the 10 points in space 50 possible positions to exist in. That might be 10^50 permutations, but it's not clear to me why you need to use every permutation for calculating Q. To me, it seems like you have 10 separate 1D coordinates, not a 10D space, unless there are some weird interaction terms with the particles?

This might make sense, but would be way clearer with a direct expression for Q.

leios replied:
```suggestion
 With that speed, it would require $$10^{41}$$ seconds on a single processor to calculate $$Q$$, which means that _even_ with all the processors in the world running in parallel (there could be billions or trillions of them), calculating $$Q$$ would still take longer than the age of the universe by many orders of magnitude!
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,273 @@
+ With that speed, it would require $$10^{41}$$ seconds on a single processor to calculate $$Q$$ - which means that _even_ with all the processors in the world running in parallel (there could be billions or trillions of them), calculating $$Q$$ would still take longer than the age of the universe - by many orders of magnitude!
+
+What's really powerful about the Metropolis approach is that you don't need to know the probability function itself - you just need a function which is _proportional_ to it. 
+What this means for the Botlzmann distribution is that you only need to know the term,
```
```suggestion
What this means for the Boltzmann distribution is that you only need to know the term,
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,273 @@
+f(\mathbf{x}) = \exp\left[{\displaystyle\frac{-E(\mathbf{x})}{T} } \right]
+$$
+
+And the Metropolis algorithm can bypass calculation of $$Q$$ altogether and use $$f(x)$$ to generate a distribution of $$x$$ which follows the probability density $$P(x)$$. 
```
This is possible because $$Q$$ is essentially just a normalization factor, so finding some other way to evaluate f and doing the normalization after-the-fact can eliminate the Q calculation entirely.

I think it's important to explicitly mention this

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,273 @@
+$$
+
+And the Metropolis algorithm can bypass calculation of $$Q$$ altogether and use $$f(x)$$ to generate a distribution of $$x$$ which follows the probability density $$P(x)$$. 
+In other words, it can sample values of $$x$$ in such away that the probability of sampling $$x$$ will follow the actual distribution $$P(x)$$ - the more likely a value of $$x$$ is according to $$P(x)$$, the more likely it will be sampled by the Metropolis algorithm. 
```
```suggestion
In other words, it can sample values of $$x$$ in such away that the probability of sampling $$x$$ will follow the actual distribution $$P(x)$$. 
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,273 @@
+Of course to be truly proportional to $$P(x)$$ the Metropolis algorithm would still take some time before it converges - but it scales very well with the dimensionality of $$\mathbf{x}$$, which in this case is proportional to the number of particles in the system, which can get very large in practice. 
+Furthermore, we often don't care about the _entire_ space of possibilities. 
+We usually care about a local region of $$P(x)$$, where the probability is relatively high, and so you would have "realistic" positions for the particles (it turns out reality tries to exist in a high probability state, who would've guessed?). 
+Because the Metropolis algorithm tends toward high probability regions, it's good at sampling those regions and providing us with a realistic collection of systems.
```
I almost feel like we could completely remove this paragraph without any loss to the chapter. Maybe the main point was a bit too confusing?

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,273 @@
+
+Finally, the Metropolis algorithm can be modified or implemented in other algorithms, and forms the basis of many advanced sampling algorithms. 
+The most popular is probably the Metropolis-Hastings algorithm {{ "hastings1970monte" | cite }} which is fundamentally the same. 
+Other algorithms that implement the algorithm are Metropolis-adjusted Langevin algorithm {{ "mala_wiki" | cite }}, and Hamiltonian Monte Carlo {{ "hmc_wiki" | cite }}, to name a few. 
```
```suggestion
Some other algorithms that use this method are Metropolis-adjusted Langevin algorithm {{ "mala_wiki" | cite }}, and Hamiltonian Monte Carlo {{ "hmc_wiki" | cite }}, to name a few. 
```

leios reviewed (commented):
On the one hand, the metropolis chapter is looking way better, but on the other hand, I am not sure I fully understand the arguments in the example application

shudipto-amin commented:
> On the one hand, the metropolis chapter is looking way better, but on the other hand, I am not sure I fully understand the arguments in the example application

Addressed your comments regarding this. Particularly, I think the difficulty of calculating $$Q$$ in the example is a bit clearer now.

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+It's intuitive for us to understand what a __discrete__ probability distribution is. 
+For example, we understand the outcomes of a coin toss very well, and also that of a dice roll. 
+For a single coin toss, we know that the probability of getting heads (H) is half, or $$P(H) = \frac{1}{2}$$. 
+Similarly, the probability of getting tails (T) is $$P(T) = \frac{1}{2}$$. 
```
```suggestion
Similarly, the probability of getting tails $$(T)$$ is $$P(T) = \frac{1}{2}$$. 
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+$$
+
+Here, $$n$$ denotes the outcome, and we used the "set notation", $$n \in\left\{H,T\right\}$$, which means "$$n$$ belongs to a set containing $$H$$ and $$T$$". 
+From the above equation, we can also assume that any other outcome for $$n$$ (such as landing on an edge) is incredibly unlikely, impossible, or simply "not allowed" (eg. just toss again if it _does_ land on its edge!).
```
Maybe "for example" instead of eg. Also, shouldn't it be e. g.?

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+For a probability distribution, it's important to take note of the set of possibilities, or the __domain__ of the distribution. 
+Here, $$\left\{H,T\right\}$$ is the domain of $$P(n)$$, telling us that $$n$$ can only be either $$H$$ or $$T$$.
+
+If we use a different system, the outcome $$n$$ could mean other things - it could also be a number. 
```
```suggestion
If we use a different system, the outcome $$n$$ could mean other things.
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+Here, $$\left\{H,T\right\}$$ is the domain of $$P(n)$$, telling us that $$n$$ can only be either $$H$$ or $$T$$.
+
+If we use a different system, the outcome $$n$$ could mean other things - it could also be a number. 
+For example, the outcome of a __die roll__ has the probability distribution,
```
```suggestion
For example, it could be a number like the outcome of a __die roll__ which has the probability distribution,
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+$$
+P(n) = \begin{matrix}
+		\displaystyle\frac 1 6 &;& n \in [1..6] 
+		\end{matrix}
```
```suggestion
		\end{matrix}.
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+		\displaystyle\frac 1 6 &;& n \in [1..6] 
+		\end{matrix}
+$$
+which is saying that the probability of $$n$$ being a whole number between 1 and 6 is $$1/6$$, and we assume that the probability of getting any other $$n$$ is 0. 
```
```suggestion
This is saying that the probability of $$n$$ being a whole number between 1 and 6 is $$1/6$$, and we assume that the probability of getting any other $$n$$ is 0. 
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+An example of a discrete probability function where the probability actually depends on $$n$$, is when $$n$$ is the sum of numbers on a __roll of two dice__. 
+In this case, $$P(n)$$  is different for each $$n$$ as some possibilities like $$n=2$$ can happen in only one possible way (by getting a 1 on both dice), whereas $$n=4$$ can happen in 3 ways (1 and 3; or 2 and 2; or 3 and 1). 
+
+The rolling two dice is a great case study for how we can construct a probability distribution, since the probability varies and it is not immediately obvious how it varies. 
```
```suggestion
The example of rolling two dice is a great case study for how we can construct a probability distribution, since the probability varies and it is not immediately obvious how it varies. 
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+So let's go ahead and construct it! 
+
+Let's first define the domain of our target $$P(n)$$. 
+We know that the lowest sum of two dice is 2 (a 1 on both dice), so $$n \geq 2$$ for sure. Similarly, the maximum is sum of two sixes, or 12, so $$n \leq 12$$ also. 
```
```suggestion
We know that the lowest sum of two dice is 2 (a 1 on both dice), so $$n \geq 2$$ for sure. Similarly, the maximum is sum of two sixes, or 12, so $$n \leq 12$$ also. 
```
```suggestion
We know that the lowest sum of two dice is 2 (a 1 on both dice), so $$n \geq 2$$ for sure. Similarly, the maximum is the sum of two sixes, or 12, so $$n \leq 12$$ also. 
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+For two dice, that is $$N = 6 \times 6 = 36$$, but we could also express it as the _sum of all frequencies_,
+
+$$
+N = \sum_n f(n)
```
```suggestion
N = \sum_n f(n),
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+So in general, if we have the function $$f(n)$$, we can get the probability as
+
+$$
+P(n) = \frac{f(n)}{\displaystyle\sum_{n} f(n)}
```
```suggestion
P(n) = \frac{f(n)}{\displaystyle\sum_{n} f(n)}.
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+P(n) = \frac{f(n)}{\displaystyle\sum_{n} f(n)}
+$$
+
+Note that $$f(n)$$ does not necessarily have to be the frequency of $$n$$ - it could be any function which is _proportional_ to $$P(n)$$, and the above definition of $$P(n)$$ would still hold. 
```
```suggestion
Note that $$f(n)$$ does not necessarily have to be the frequency of $$n$$ &ndash; it could be any function which is _proportional_ to $$P(n)$$, and the above definition of $$P(n)$$ would still hold. 
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+$$
+
+Note that $$f(n)$$ does not necessarily have to be the frequency of $$n$$ - it could be any function which is _proportional_ to $$P(n)$$, and the above definition of $$P(n)$$ would still hold. 
+And it's easy to check that the sum is now equal to 1, since
```
```suggestion
It's easy to check that the sum is now equal to 1, since
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+And it's easy to check that the sum is now equal to 1, since
+
+$$
+\sum_n P(n) = \frac{\displaystyle\sum_{n}f(n)}{\displaystyle\sum_{n} f(n)} = 1
```
```suggestion
\sum_n P(n) = \frac{\displaystyle\sum_{n}f(n)}{\displaystyle\sum_{n} f(n)} = 1.
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+And to calculate it, we simply have to sum up all the probabilities for each value of $$n$$ in that range, i.e.,
+
+$$
+\mathbb{P}(a \leq n \leq b) = \sum_{n=a}^{b} P(n)
```
```suggestion
\mathbb{P}(a \leq n \leq b) = \sum_{n=a}^{b} P(n).
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+## Probability Density Functions
+
+What if instead of a discrete variable $$n$$, we had a continuous variable $$x$$, like temperature or weight? 
+In that case, it doesn't make sense to ask what the probability is of $$x$$ being _exactly_ a particular number - there are infinite possible real numbers, after all, so the probability of $$x$$ being exactly any one of them is essentially zero! 
```
```suggestion
In that case, it doesn't make sense to ask what the probability is of $$x$$ being _exactly_ a particular number &ndash; there are infinite possible real numbers, after all, so the probability of $$x$$ being exactly any one of them is essentially zero! 
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+Now, if we are interested in the probability of the range of values that lie between $$x_0$$ and $$x_0 + dx$$, all we have to do is calculate the _area_ of the green sliver above. 
+This is the defining feature of a  probability density function: 
+
+> the probability of a range of values is the _area_ of the region under the probability density curve which is within that range. 
```
I feel like the quote here makes it less visible and also make me feel like it is coming from another source. Maybe bold it instead?

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+So if $$dx$$ is infinitesimally small, then the area of the green sliver becomes $$P(x)dx$$, and hence,
+
+$$
+\mathbb{P}(x_0 \leq x \leq x_0 + dx) = P(x)dx
```
```suggestion
\mathbb{P}(x_0 \leq x \leq x_0 + dx) = P(x)dx.
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+Thus, to obtain the probability of $$x$$ lying within a range, we simply integrate $$P(x)$$ between that range, i.e.,
+
+$$
+\mathbb{P}(a \leq x \leq b ) = \int_a^b P(x)dx
```
```suggestion
\mathbb{P}(a \leq x \leq b ) = \int_a^b P(x)dx.
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+This is analagous to finding the probability of a range of discrete values from the previous section:
+
+$$
+\mathbb{P}(a \leq n \leq b) = \sum_{n=a}^{b} P(n)
```
```suggestion
\mathbb{P}(a \leq n \leq b) = \sum_{n=a}^{b} P(n).
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+\mathbb{P}(a \leq n \leq b) = \sum_{n=a}^{b} P(n)
+$$
+
+And the fact that all probabilities must sum to 1 translates to
```
```suggestion
The fact that all probabilities must sum to 1 translates to
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+And the fact that all probabilities must sum to 1 translates to
+
+$$
+\int_D P(x)dx = 1
```
```suggestion
\int_D P(x)dx = 1.
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+We can get the probability density function by normalizing it in a similar way, except that we integrate instead of sum:
+
+$$
+P(\mathbf{x}) = \frac{f(\mathbf{x})}{\int_D f(\mathbf{x})d\mathbf{x}}
```
```suggestion
P(\mathbf{x}) = \frac{f(\mathbf{x})}{\int_D f(\mathbf{x})d\mathbf{x}}.
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+We first integrate it (or do a quick google search, as it is rather tricky) to get
+
+$$
+N = \int_{-\infty}^{\infty} e^{-x^2} dx = \sqrt{\pi}
```
```suggestion
N = \int_{-\infty}^{\infty} e^{-x^2} dx = \sqrt{\pi}.
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+N = \int_{-\infty}^{\infty} e^{-x^2} dx = \sqrt{\pi}
+$$
+
+and so we have a Gaussian probability distribution,
```
```suggestion
Now we have a Gaussian probability distribution,
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+and so we have a Gaussian probability distribution,
+
+$$
+P(x) = \frac{1}{N} e^{-x^2} = \frac{1}{\sqrt{\pi}} e^{-x^2}
```
```suggestion
P(x) = \frac{1}{N} e^{-x^2} = \frac{1}{\sqrt{\pi}} e^{-x^2}.
```

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+$$
+
+In general, normalization can allow us to create a probability distribution out of almost any function $$f(x)$$. 
+There are really only two rules that $$f(\mathbf{x})$$ must satisfy to be a candidate for a probability density distribution:
```
Why are we bolding x now?

shudipto-amin replied:
I moved from specific 1D example to talking about general formulation, so x is bold to represent multidimensional coordinates.

leios commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+In general, normalization can allow us to create a probability distribution out of almost any function $$f(x)$$. 
+There are really only two rules that $$f(\mathbf{x})$$ must satisfy to be a candidate for a probability density distribution:
+1. The integral of $$f(\mathbf{x})$$ over any subset of $$D$$ (denoted by $$S$$) has to be non-negative (it can be zero):
+$$
+\int_{S}f(\mathbf{x})d\mathbf{x} \geq 0
+$$ 
+2. The following integral must be finite:
+$$
+\int_{D} f(\mathbf{x})d\mathbf{x}
+$$ 
```
```suggestion
$$
\int_{S}f(\mathbf{x})d\mathbf{x} \geq 0.
$$ 
2. The following integral must be finite:
$$
\int_{D} f(\mathbf{x})d\mathbf{x}.
$$ 
```

leios reviewed (commented):
I am happy with the probability density section and think it's pretty much ready to go. Now to take a closer look at metropilis.

Thanks again!

shudipto-amin commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+Now, if we are interested in the probability of the range of values that lie between $$x_0$$ and $$x_0 + dx$$, all we have to do is calculate the _area_ of the green sliver above. 
+This is the defining feature of a  probability density function: 
+
+> the probability of a range of values is the _area_ of the region under the probability density curve which is within that range. 
```
```suggestion
 __the probability of a range of values is the _area_ of the region under the probability density curve which is within that range.__ 
```

shudipto-amin commented on contents/probability/distributions/distributions.md:
```diff
@@ -0,0 +1,215 @@
+P(x) = \frac{1}{N} e^{-x^2} = \frac{1}{\sqrt{\pi}} e^{-x^2}
+$$
+
+In general, normalization can allow us to create a probability distribution out of almost any function $$f(x)$$. 
```
```suggestion
In general, normalization can allow us to create a probability distribution out of almost any function $$f(\mathbf{x})$$. 
```

Amaras commented:
I need to tell you two things @shudipto-amin.

First we receive a notification on the discord server after every single commit.
This gets overwhelming very quickly. So could you please commit suggestions in batches in the future? This applies to this chapter review and all future reviews.

Second, I have a few problems with your notation, since I come from a mathematical background, the main thing being that I have never seen the [2..12] notation used anywhere. I have already seen "double brackets" (something that looks like I[2, 12]I, I can't reproduce it better right now) and $ [2, 12] \cap \mathbb{N}$ for the integer interval [2, 12].
If you could point me to several mathematical sources that use your notation, I'll gladly read them and maybe even use it myself, but for now, it's not something I recognise easily.

shudipto-amin commented:
@Amaras 
> I need to tell you two things @shudipto-amin.
> 
> First we receive a notification on the discord server after every single commit. This gets overwhelming very quickly. So could you please commit suggestions in batches in the future? This applies to this chapter review and all future reviews.
> 
I'm sorry about that, I was adding from conversations where that seems to be the only option. Later, I did discover that option in the files tab, and started batching about half-way through. 

> Second, I have a few problems with your notation, since I come from a mathematical background, the main thing being that I have never seen the [2..12] notation used anywhere. I have already seen "double brackets" (something that looks like I[2, 12]I, I can't reproduce it better right now) and $ [2, 12] \cap \mathbb{N}$ for the integer interval [2, 12]. If you could point me to several mathematical sources that use your notation, I'll gladly read them and maybe even use it myself, but for now, it's not something I recognise easily.
I personally only discovered this notation recently - I normally see something like your example with $\mathbb{N}$. However, I thought this would be notationally simpler, since I explain it anyway. I also thought it would be a bit more familiar to people with some coding background, but I'm not really sure. The only public source I can remember right now is this [Wikipedia page](https://en.wikipedia.org/wiki/Interval_(mathematics)#Integer_intervals). If you think I should change it, I don't mind.

Amaras commented:
> I'm sorry about that, I was adding from conversations where that seems to be the only option. Later, I did discover that option in the files tab, and started batching about half-way through.

No problem, but keep that in mind for the future.

> I personally only discovered this notation recently - I normally see something like your example with $\mathbb{N}$. However, I thought this would be notationally simpler, since I explain it anyway. I also thought it would be a bit more familiar to people with some coding background, but I'm not really sure. The only public source I can remember right now is this [Wikipedia page](https://en.wikipedia.org/wiki/Interval_(mathematics)#Integer_intervals). If you think I should change it, I don't mind.

Yeah, that just tells me we need a mathematical notation chapter. I personally would prefer the ⟦a, b⟧ notation, but that's not a native latex symbol if I remember correctly.

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,280 @@
+$$
+
+where $$x_i, y_i, z_i$$ are the spatial coordinates of particle $$i$$. 
+So altogether there are $$3N$$ coordinates - making $$E(\mathbf{x})$$ a $$3N$$ dimensional function, which can be a computationally intensive calculation on it's own. But it doesn't end there!
```
```suggestion
So altogether there are $$3N$$ coordinates &ndash; making $$E(\mathbf{x})$$ a $$3N$$ dimensional function, which can be a computationally intensive calculation on it's own. But it doesn't end there!
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,280 @@
+Let's assume that the particles _interact_, meaning that the position of one particle affects that of another. 
+This could be the case, for example, if all the particles were charged, and so they would be repelling or attracting each other. 
+This means that the energy $$E(\mathbf{x}) = E(x_1,...,x_{10})$$ of the system is a 10D function, and it would not be possible to simplify it any further due to the interactions. 
+Thus, the Boltzmann factor, $$\exp\left[-E(\mathbf{x})/T\right]$$, is also a 10D function. And so, to calculate $$Q$$, we would have to integrate the Boltzmann factor 10 times, one for each coordinate,
```
```suggestion
Thus, the Boltzmann factor, $$\exp\left[-E(\mathbf{x})/T\right]$$, is also a 10D function. To calculate $$Q$$, we would have to integrate the Boltzmann factor 10 times, one for each coordinate,
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,280 @@
+In most cases, there is no known analytical expression for the above integral, so it has to be done numerically.
+To do so, imagine that we divide the 1D line segment into only 50 different intervals, allowing each particle to take on 50 different positions. 
+This is equivalent to dividing the length of a football field into intervals of about 2 meters - not a resolution you'd wanna watch a game in! 
+Even with such poor resolution, the number of different combinations of positions is $$10^{50}$$ - a colossal number indeed. 
```
```suggestion
Even with such poor resolution, the number of different combinations of positions is $$10^{50}$$ &ndash; a colossal number indeed. 
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,280 @@
+To see how large this number is, imagine that a single computation of $$E(\mathbf{x})$$ took only 1 nanosecond on a single processor, which is much faster than most energy calculations for physical systems in practice.
+ With that speed, it would require $$10^{41}$$ seconds on a single processor to calculate $$Q$$ - which means that _even_ with all the processors in the world running in parallel (there could be billions or trillions of them), calculating $$Q$$ would still take longer than the age of the universe - by many orders of magnitude!
+
+What's really powerful about the Metropolis approach is that you don't need to know the probability function itself - you just need a function which is _proportional_ to it. 
```
```suggestion
What's really powerful about the Metropolis approach is that you don't need to know the probability function itself.
Instead, you just need a function which is _proportional_ to it. 
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,280 @@
+f(\mathbf{x}) = \exp\left[{\displaystyle\frac{-E(\mathbf{x})}{T} } \right]
+$$
+
+The Metropolis algorithm can bypass calculation of $$Q$$ altogether and use $$f(x)$$ to generate a distribution of $$x$$ which follows the probability density $$P(x)$$. 
```
```suggestion
The Metropolis algorithm can bypass the calculation of $$Q$$ altogether and use $$f(x)$$ to generate a distribution of $$x$$ which follows the probability density $$P(x)$$. 
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,280 @@
+Numerical normalization can then be done by using the total number of samples instead of performing an integration. 
+This fact dramatically reduces the number of calculations needed to approximate the probability distribution.
+
+Finally, the Metropolis algorithm can be modified or implemented in other algorithms, and forms the basis of many advanced sampling algorithms. 
```
```suggestion
Finally, the Metropolis algorithm can be modified or implemented in other methods, and forms the basis of many advanced sampling algorithms. 
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,280 @@
+The Metropolis algorithm {{ "metropolis1953equation" | cite }} is a slightly more advanced Monte Carlo method which uses random numbers to approximate a [probability distribution](../probability/distributions/distributions.md):
+
+$$
+P(\mathbf{x}) = \frac{f(\mathbf{x})}{\displaystyle\int_D f(\mathbf{x})d\mathbf{x}}
```
```suggestion
P(\mathbf{x}) = \frac{f(\mathbf{x})}{\displaystyle\int_D f(\mathbf{x})d\mathbf{x}},
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,280 @@
+and $$f(\mathbf{x})$$ is some a function that is proportional to $$P(x)$$, such as a statistical frequency distribution, i.e., one that counts the number of occurences of each $$\mathbf{x}$$.
+The integral in the denominator is the __normalization factor__ which ensures that the sum of all probabilities is unity, i.e.,
+$$
+\int_D P(\mathbf{x}) d\mathbf{x} = 1
```
```suggestion
\int_D P(\mathbf{x}) d\mathbf{x} = 1.
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,280 @@
+A one-dimensional example is the __normal distribution__, or __Gaussian distribution__, given by
+
+$$
+P(x) = \frac{e^{-x^2}}{\displaystyle\int_{-\infty}^{\infty} e^{-x^2} dx} = \frac{1}{\sqrt{\pi}} e^{-x^2}
```
```suggestion
P(x) = \frac{e^{-x^2}}{\displaystyle\int_{-\infty}^{\infty} e^{-x^2} dx} = \frac{1}{\sqrt{\pi}} e^{-x^2}.
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,280 @@
+For such systems, we can usually describe the __potential energy__ {{ "potential_energy_wiki" | cite }} of the system as a function of the coordinates of all particles, $$\mathbf{x}$$,
+
+$$
+E(\mathbf{x}) = E(x_1, y_1, z_1, x_2, y_2, z_2, ... ,x_N, y_N, z_N) 
```
```suggestion
E(\mathbf{x}) = E(x_1, y_1, z_1, x_2, y_2, z_2, ... ,x_N, y_N, z_N),
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,280 @@
+The physicist Ludwig Boltzmann {{ "ludwig_boltzmann_wiki" | cite }} discovered that when such a system is in equilibrium at some temperature $$T$$, you can describe the probability density of the system for any set of coordinates $$\mathbf{x}$$ using, {{ "boltzmann_distribution_wiki" | cite }} 
+
+$$
+P(\mathbf{x}) = \frac{\displaystyle \exp\left[{\displaystyle\frac{-E(\mathbf{x})}{T} } \right]} {Q}
```
```suggestion
P(\mathbf{x}) = \frac{\displaystyle \exp\left[{\displaystyle\frac{-E(\mathbf{x})}{T} } \right]} {Q},
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,280 @@
+where the numerator is called the __Boltzmann factor__, and $$Q$$ is the [normalization constant](../probability/distributions/distributions.md),
+
+$$
+Q = \int_D \exp\left[{\displaystyle\frac{-E(\mathbf{x})}{T} } \right] d\mathbf{x}
```
```suggestion
Q = \int_D \exp\left[{\displaystyle\frac{-E(\mathbf{x})}{T} } \right] d\mathbf{x}.
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,280 @@
+Thus, the Boltzmann factor, $$\exp\left[-E(\mathbf{x})/T\right]$$, is also a 10D function. And so, to calculate $$Q$$, we would have to integrate the Boltzmann factor 10 times, one for each coordinate,
+
+$$
+Q = \int_{x_1} \dots \int_{x_{10}} \exp\left[\frac{-E(x_1,\dots x_{10})}{T}\right]\ dx_1\dots dx_{10}
```
```suggestion
Q = \int_{x_1} \dots \int_{x_{10}} \exp\left[\frac{-E(x_1,\dots x_{10})}{T}\right]\ dx_1\dots dx_{10}.
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,280 @@
+What this means for the Boltzmann distribution is that you only need to know the term,
+
+$$
+f(\mathbf{x}) = \exp\left[{\displaystyle\frac{-E(\mathbf{x})}{T} } \right]
```
```suggestion
f(\mathbf{x}) = \exp\left[{\displaystyle\frac{-E(\mathbf{x})}{T} } \right].
```

leios reviewed (commented):
Looks good! Mostly done, just have to look at the second half of Metropolis now!

Amaras commented:
> @Amaras, I used this negative space hack, and it actually looks alright. Does this work?
Yeah, it looks good :)

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,281 @@
+## A Random Walk in One Dimension
+
+In the rest of this chapter, we will look at $$1$$D examples to understand the Metropolis algorithm. 
+Although the algorithm is not particularly efficient in just one dimension, it is much easier to understand and learn how to implement than in higher dimensions. 
```
```suggestion
Although the algorithm is not particularly efficient in just one dimension, it is much easier to understand in one dimension than in multiple dimensions. 
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,281 @@
+This is because there is a long "barrier" of low probability region between the third peak and second peak. 
+This may not necessarily be a bad thing &ndash; sometimes one might want to estimate how long something takes to transition from one state to another, and often these peaks represent such 'states'. 
+So averaging over many metropolis runs may give some estimate of these transition times. 
+And if global sampling is the goal, the process of exploration could be sped up by choosing larger step sizes for the walker, for example by choosing step size $$g$$ from an interval like $$(-3,3)$$ instead of $$(-1,1)$$. 
```
```suggestion
If global sampling is the goal, the process of exploration could be sped up by choosing larger step sizes for the walker, for example by choosing step size $$g$$ from an interval like $$(-3,3)$$ instead of $$(-1,1)$$. 
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,281 @@
+
+
+## The Algorithm for a One Dimensional Example
+
```
There should be a comment here introducing the upcoming sections

shudipto-amin replied:
Couldn't come up with anything other than the line (147) below. Feel free to throw in some suggestions.

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,281 @@
+P(x) = \frac{f(x)}{\int_{-10}^{10} f(x)},
+$$
+
+where $$f(x)$$ is the function we know and is given by
```
```suggestion
where $$f(x)$$ is the same function we have shown above and is given by
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,281 @@
+{% endmethod %}
+
+However, $$g$$ can be any function symmetric about $$0$$ for the above algorithm to work. 
+For example, it can be a number chosen randomly from a discrete list, such as $$[ -3, -1, -1, +1, +1, +3]$$. 
```
To be clear: this was meant to have multiple 1's, right?

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,281 @@
+However, $$g$$ can be any function symmetric about $$0$$ for the above algorithm to work. 
+For example, it can be a number chosen randomly from a discrete list, such as $$[ -3, -1, -1, +1, +1, +3]$$. 
+It can also be a number chosen from a symmetric continuos distribution, like the Gaussian, $$e^{-x^2}$$. 
+In higher dimensions, the function should be spherically symmetric, such as multidimensional Gaussian function, $$e^{-(x^2 +y^2 + ...)}$$. 
```
```suggestion
In higher dimensions, the function should be spherically symmetric, such as a multidimensional Gaussian function, $$e^{-(x^2 +y^2 + ...)}$$. 
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,281 @@
+
+
+## Example Code
+The following code puts everything together, and runs Metropolis algorithm for a number of  steps given by `num_steps`. 
```
```suggestion
The following code puts everything together, and runs the Metropolis algorithm for a number of  steps given by `num_steps`. 
```

leios reviewed (commented):
Overall, very small changes to the text. Sorry to keep splitting this up into a bunch of smaller reviews.

I'll look at the code tomorrow and then I think the PR will be mostly ready to go!

shudipto-amin reviewed (commented):
Added a line to introduce subsections in " The Algorithm for a One D..." section. Not sure if it's sufficient. Addressed all other minor edits in a previous commit.

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,283 @@
+P(\mathbf{x}) = \frac{f(\mathbf{x})}{\displaystyle\int_D f(\mathbf{x})d\mathbf{x}},
+$$
+
+where $$D$$ is the domain of $$P(\mathbf{x})$$, i.e., all possible values of the $$\mathbf{x}$$ for which $$P(\mathbf{x})$$ is defined; 
```
```suggestion
where $$D$$ is the domain of $$P(\mathbf{x})$$, i.e., all possible values of the $$\mathbf{x}$$ for which $$P(\mathbf{x})$$ is defined.
```

I think this is a long run-on sentence

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,283 @@
+$$
+
+where $$D$$ is the domain of $$P(\mathbf{x})$$, i.e., all possible values of the $$\mathbf{x}$$ for which $$P(\mathbf{x})$$ is defined; 
+and $$f(\mathbf{x})$$ is some a function that is proportional to $$P(x)$$, such as a statistical frequency distribution, i.e., one that counts the number of occurences of each $$\mathbf{x}$$.
```
```suggestion
$$f(\mathbf{x})$$ is a function that is proportional to $$P(x)$$, such as a statistical frequency distribution which counts the number of occurences of each $$\mathbf{x}$$.
```

lots of i.e.s

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,283 @@
+
+Although convergence occurs eventually, not all parts of the distribution achieve convergence quickly. 
+Note from the animation above, that the walker very quickly replicates the distribution of the two peaks on the left, but takes quite a while to even reach the third peak to the right. 
+This is because there is a long "barrier" of low probability region between the third peak and second peak. 
```
```suggestion
This is because there is a long low probability region between the third peak and second peak that acts as a "barrier." 
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,283 @@
+On the other hand, a walker with a large step size may not sample nearby regions accurately &ndash; and actually has a higher chance of being rejected if the walker is already in a high probability region, since the acceptance ratio is more drastic for large steps. 
+The effect of step-size on the walker's efficiency is far from obvious! 
+
+How to choose $$g$$ is in itself a research field and depends on what the goal of the sampling is. 
```
```suggestion
How to choose $$g$$ is in itself a research field and depends on what the goal of the sampling is. 
```
```suggestion
How to choose $$g$$ is, itself, a research field and depends on what the goal of the sampling is. 
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,283 @@
+The effect of step-size on the walker's efficiency is far from obvious! 
+
+How to choose $$g$$ is in itself a research field and depends on what the goal of the sampling is. 
+Some methods even use an "adaptive" method where $$g$$ is "trained" on-the-fly using a learning algorithm! 
```
```suggestion
Some techniques even use an "adaptive" method where $$g$$ is "trained" on-the-fly using a learning algorithm! 
```

leios commented on contents/metropolis/code/python/metropolis.py:
```diff
@@ -0,0 +1,90 @@
+        
+ 
+if __name__ == "__main__":
+    xmin, xmax = -7, 7
```
Didn't we say in the chapter that our domain was from -10 to 10?

shudipto-amin replied:
```suggestion
    xmin, xmax = -10, 10
```

leios commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,283 @@
+3. Accept proposal, $$x'$$ with probability $$A$$. If your programming language doesn't have a built-in method for this,
+	* Generate a random number $$u$$ between $$0$$ and $$1$$.
+    * If $$ u \leq A $$, then __accept__ move, and set new position, $$x_{t+1} = x' $$.
+    * Otherwise, __reject__ move, and set new position to current, $$x_{t+1} = x_t $$.
```
```suggestion
    * Otherwise, __reject__ move, and set new position to current position, $$x_{t+1} = x_t $$.
```

leios reviewed (approved):
Overall, this looks good to me. There are just a few small changes to be made, so I am happy to go ahead and approve this! Good work and thanks for the patience!

shudipto-amin commented on contents/metropolis/metropolis.md:
```diff
@@ -0,0 +1,283 @@
+On the other hand, a walker with a large step size may not sample nearby regions accurately &ndash; and actually has a higher chance of being rejected if the walker is already in a high probability region, since the acceptance ratio is more drastic for large steps. 
+The effect of step-size on the walker's efficiency is far from obvious! 
+
+How to choose $$g$$ is in itself a research field and depends on what the goal of the sampling is. 
```
```suggestion
The question of how to choose an optimal $$g$$ is a research area on its own, and depends largely on what the goal of the sampling is.
```

shudipto-amin reviewed (commented):
Suggesting changes to address remaining comments by Leios, namely:
1. Changing the domain of x in python code,
2. Rewording the statement regarding finding an optimal $$g$$.
